File: Data/analysis_pipeline.py
"""
Core analysis pipeline for SSP biomechanics data.
=================================================

This module provides the core analysis functions for processing
biomechanical test data from the supraspinatus tear study.

Key functions:
- classify_sample: Parse filenames to determine sample groups
- find_best_stiffness: Find the stiffest linear region in load-displacement curves
- process_all_files: Batch process all data files
- run_pipeline: Execute the complete analysis pipeline
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

from ssp_config import get_config_paths, get_group_ids
from utils import (
    classify_sample,
    safe_trapezoid,
    validate_raw_data,
    normalize_load_column,
)

# Configure module logger
logger = logging.getLogger(__name__)

# Re-export classify_sample for backward compatibility
__all__ = [
    "classify_sample",
    "find_best_stiffness",
    "process_all_files",
    "generate_statistics",
    "save_results",
    "create_plots",
    "run_pipeline",
]


def find_best_stiffness(
    x: np.ndarray,
    y: np.ndarray,
    window_size: int,
    r2_threshold: float
) -> Tuple[float, float, float, int, int]:
    """
    Scan the curve to find the stiffest linear region.
    
    Uses a sliding window approach to find the region with the highest
    stiffness (slope) while maintaining good linearity (R² above threshold).
    
    Args:
        x: Displacement values (mm)
        y: Load values (N)
        window_size: Number of points in the sliding window
        r2_threshold: Minimum R² value to consider a region as linear
        
    Returns:
        Tuple of (slope, intercept, r_squared, start_idx, end_idx)
        Returns (nan, nan, nan, 0, 0) if no valid region is found.
    """
    best_r2 = -np.inf
    best_slope = 0.0
    best_params: Tuple[float, float, float, int, int] = (np.nan, np.nan, np.nan, 0, 0)

    x = np.asarray(x, dtype=np.float64)
    y = np.asarray(y, dtype=np.float64)

    n_points = len(x)
    if n_points < window_size:
        logger.warning(
            f"Insufficient data points ({n_points}) for window size ({window_size})"
        )
        return best_params

    for i in range(n_points - window_size):
        x_win = x[i : i + window_size].reshape(-1, 1)
        y_win = y[i : i + window_size]

        model = LinearRegression().fit(x_win, y_win)
        y_pred = model.predict(x_win)

        r2 = r2_score(y_win, y_pred)
        slope = float(model.coef_[0])
        intercept = float(model.intercept_)

        # Prefer regions with R² above threshold and highest slope
        if r2 >= r2_threshold:
            if slope > best_slope:
                best_slope = slope
                best_params = (slope, intercept, r2, i, i + window_size)

        # Fallback: if no region meets threshold, use best R²
        if r2 > best_r2 and best_slope == 0:
            best_r2 = r2
            best_params = (slope, intercept, r2, i, i + window_size)

    return best_params


def process_all_files(
    data_dir: Path,
    tfl_ids: List[str],
    msc_ids: List[str],
    r2_threshold: float,
    window_fraction: float,
    min_window: int
) -> Optional[pd.DataFrame]:
    """
    Process all CSV files in the data folder.
    
    Args:
        data_dir: Path to directory containing CSV data files
        tfl_ids: List of subject IDs in the TFL treatment group
        msc_ids: List of subject IDs in the MSC treatment group
        r2_threshold: Minimum R² for linear region detection
        window_fraction: Fraction of data points for sliding window
        min_window: Minimum window size in data points
        
    Returns:
        DataFrame with analysis results, or None if processing fails.
    """
    logger.info("=" * 60)
    logger.info("SSP BIOMECHANICS ANALYSIS PIPELINE")
    logger.info("=" * 60)

    if not data_dir.exists():
        logger.error(f"Folder not found: {data_dir}")
        return None

    files = sorted([f for f in data_dir.iterdir() if f.suffix.lower() == ".csv"])
    data_records: List[Dict] = []

    logger.info(f"Processing {len(files)} files...")

    for file_path in files:
        filename = file_path.name
        try:
            s_id, cond, sub = classify_sample(filename, tfl_ids, msc_ids)

            if cond == "Unknown":
                logger.warning(f"Skipping {filename}: Could not determine NO/OPER.")
                continue

            df = pd.read_csv(file_path)
            
            # Validate data
            validation_errors = validate_raw_data(df)
            if validation_errors:
                logger.warning(f"Skipping {filename}: {'; '.join(validation_errors)}")
                continue
            
            # Normalize load column
            df = normalize_load_column(df)
            y_col = "LoadN"

            # Truncate at max load (failure point)
            max_idx = df[y_col].idxmax()
            df_trunc = df.iloc[: max_idx + 1].copy()
            x = df_trunc["Crossheadmm"].values
            y = df_trunc[y_col].values

            # Calculate energy (area under curve)
            energy_mJ = safe_trapezoid(y, x)

            # Find stiffness using sliding window
            window_span = max(min_window, int(len(x) * window_fraction))
            slope, intercept, r2, idx_start, idx_end = find_best_stiffness(
                x, y, window_span, r2_threshold
            )
            stiffness_N_mm = slope

            data_records.append({
                "Filename": filename,
                "SampleID": s_id,
                "Subgroup": sub,
                "MaxLoad_N": float(df[y_col].max()),
                "Stiffness_N_mm": stiffness_N_mm,
                "Energy_mJ": energy_mJ,
                "R2_Score": r2,
                "Linear_Start_Idx": idx_start,
                "Linear_End_Idx": idx_end
            })

            logger.info(
                f"[OK] {filename} -> {sub}: "
                f"MaxLoad={df[y_col].max():.1f}N, Stiffness={stiffness_N_mm:.1f}N/mm"
            )

        except pd.errors.EmptyDataError:
            logger.error(f"Empty data file: {filename}")
        except pd.errors.ParserError as e:
            logger.error(f"CSV parsing error in {filename}: {e}")
        except KeyError as e:
            logger.error(f"Missing column in {filename}: {e}")
        except Exception as e:
            logger.error(f"Unexpected error processing {filename}: {e}", exc_info=True)

    metadata = pd.DataFrame(data_records)
    logger.info(f"Processed {len(metadata)} samples successfully.")

    return metadata


def generate_statistics(metadata: pd.DataFrame) -> pd.DataFrame:
    """
    Generate group statistics from the analysis results.
    
    Args:
        metadata: DataFrame with analysis results from process_all_files()
        
    Returns:
        DataFrame with summary statistics grouped by Subgroup.
    """
    logger.info("=" * 60)
    logger.info("GROUP STATISTICS")
    logger.info("=" * 60)

    def list_ids(series: pd.Series) -> str:
        return ", ".join(sorted(series.unique()))

    stats = metadata.groupby("Subgroup").agg(
        MaxLoad_Mean=("MaxLoad_N", "mean"),
        MaxLoad_Std=("MaxLoad_N", "std"),
        Stiffness_Mean=("Stiffness_N_mm", "mean"),
        Stiffness_Std=("Stiffness_N_mm", "std"),
        Energy_Mean=("Energy_mJ", "mean"),
        Energy_Std=("Energy_mJ", "std"),
        Count=("SampleID", "count"),
        Sample_List=("SampleID", list_ids)
    ).round(2)

    logger.info("\n" + stats.to_string())

    return stats


def save_results(
    metadata: pd.DataFrame,
    stats: pd.DataFrame,
    results_dir: Path
) -> Tuple[Path, Path]:
    """
    Save analysis results to CSV files.
    
    Args:
        metadata: DataFrame with detailed sample results
        stats: DataFrame with group statistics
        results_dir: Directory to save output files
        
    Returns:
        Tuple of (detail_path, stats_path) for the saved files.
    """
    logger.info("=" * 60)
    logger.info("SAVING RESULTS")
    logger.info("=" * 60)

    results_dir.mkdir(parents=True, exist_ok=True)

    detail_path = results_dir / "Experiment_Master_Log_Detailed.csv"
    metadata.to_csv(detail_path, index=False)
    logger.info(f"Saved: {detail_path}")

    stats_path = results_dir / "Group_Statistics_Detailed.csv"
    stats.to_csv(stats_path)
    logger.info(f"Saved: {stats_path}")
    
    return detail_path, stats_path


def create_plots(metadata: pd.DataFrame, results_dir: Path) -> Path:
    """
    Create summary bar plots for each treatment group.
    
    Args:
        metadata: DataFrame with analysis results
        results_dir: Directory to save the plot
        
    Returns:
        Path to the saved plot file.
    """
    logger.info("=" * 60)
    logger.info("GENERATING PLOTS")
    logger.info("=" * 60)

    import matplotlib.pyplot as plt

    # Use a style that's available in most matplotlib versions
    try:
        plt.style.use("seaborn-v0_8-whitegrid")
    except OSError:
        try:
            plt.style.use("seaborn-whitegrid")
        except OSError:
            logger.debug("Seaborn style not available, using default")
    
    plt.rcParams["figure.figsize"] = [10, 6]

    fig, axes = plt.subplots(1, 3, figsize=(15, 5))

    groups = ["NON", "TFL", "MSC"]
    colors = ["#2ecc71", "#3498db", "#e74c3c"]

    for i, metric in enumerate(["MaxLoad_N", "Stiffness_N_mm", "Energy_mJ"]):
        means = [metadata[metadata["Subgroup"] == g][metric].mean() for g in groups]
        stds = [metadata[metadata["Subgroup"] == g][metric].std() for g in groups]
        counts = [len(metadata[metadata["Subgroup"] == g]) for g in groups]

        bars = axes[i].bar(groups, means, color=colors, alpha=0.8, edgecolor="black")
        axes[i].errorbar(groups, means, yerr=stds, fmt="none", color="black", capsize=5)

        for j, bar in enumerate(bars):
            std_val = stds[j] if not np.isnan(stds[j]) else 0
            axes[i].text(
                bar.get_x() + bar.get_width() / 2,
                bar.get_height() + std_val + 5,
                f"n={counts[j]}",
                ha="center",
                va="bottom",
                fontsize=10
            )

        titles = {
            "MaxLoad_N": "Max Load (N)",
            "Stiffness_N_mm": "Stiffness (N/mm)",
            "Energy_mJ": "Energy (mJ)"
        }
        axes[i].set_title(titles[metric], fontsize=14, fontweight="bold")
        axes[i].set_ylabel(titles[metric])
        axes[i].set_xlabel("Treatment Group")

    plt.tight_layout()

    plot_path = results_dir / "Combined_Group_Plots.png"
    plt.savefig(plot_path, dpi=150, bbox_inches="tight")
    plt.close()

    logger.info(f"Saved: {plot_path}")
    return plot_path


def run_pipeline(config: Dict, data_root: Path) -> Optional[pd.DataFrame]:
    """
    Run the full analysis pipeline.
    
    This is the main entry point that orchestrates the entire analysis:
    1. Load configuration
    2. Process all data files
    3. Generate statistics
    4. Save results and create plots
    
    Args:
        config: Configuration dictionary (from config.json)
        data_root: Root directory for data files
        
    Returns:
        DataFrame with analysis results, or None if processing fails.
    """
    # Configure logging for pipeline run
    logging.basicConfig(
        level=logging.INFO,
        format='%(message)s'
    )
    
    paths = get_config_paths(config, data_root)
    data_dir = paths["selected_data_dir"]
    results_dir = paths["results_dir"]

    tfl_ids, msc_ids = get_group_ids(config)

    analysis_cfg = config.get("analysis", {})
    r2_threshold = float(analysis_cfg.get("stiffness_r2_threshold", 0.99))
    window_fraction = float(analysis_cfg.get("stiffness_window_fraction", 0.1))
    min_window = int(analysis_cfg.get("stiffness_min_window", 5))

    metadata = process_all_files(
        data_dir=data_dir,
        tfl_ids=tfl_ids,
        msc_ids=msc_ids,
        r2_threshold=r2_threshold,
        window_fraction=window_fraction,
        min_window=min_window
    )

    if metadata is None or len(metadata) == 0:
        logger.error("No data processed. Check your data folder.")
        return None

    stats = generate_statistics(metadata)
    save_results(metadata, stats, results_dir)
    create_plots(metadata, results_dir)

    logger.info("=" * 60)
    logger.info("[OK] ANALYSIS COMPLETE!")
    logger.info("=" * 60)
    logger.info(f"Total samples processed: {len(metadata)}")
    logger.info(f"Results saved to: {results_dir}")
    logger.info("Files generated:")
    logger.info("  - Experiment_Master_Log_Detailed.csv")
    logger.info("  - Group_Statistics_Detailed.csv")
    logger.info("  - Combined_Group_Plots.png")
    
    return metadata

File: Analysis_Pipeline.m
%% SSP Biomechanics Analysis Pipeline
% Comparative evaluation of surgical methods in the treatment of chronic
% irreparable supraspinatus tears: a biomechanical study
%
% Authors: E. Koumantou, C. J. Feroussis, E. D. Pasiou, S. K. Kourkoulis
%
% This script processes load-displacement data from biomechanical tests
% and calculates key metrics: Max Load, Stiffness, and Energy.

%% Configuration
clear; clc; close all;

% --- PATHS ---
DATA_FOLDER = "C:\Users\elena\OneDrive - Imperial College London\2_NTUA\2_SSP_Project\Data\Selected_data";
RESULTS_FOLDER = "C:\Users\elena\OneDrive - Imperial College London\2_NTUA\2_SSP_Project\Data\Results";

% Create Results folder if it doesn't exist
if ~exist(RESULTS_FOLDER, 'dir')
    mkdir(RESULTS_FOLDER);
end

% --- SUBGROUP ASSIGNMENTS ---
% These define which OPERATED samples belong to each treatment group.
% Non-operated samples are automatically assigned to 'NON' (control) group.
TFL_IDS = {'B1', 'B10', 'B11', 'C1', 'C2', 'D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8'};
MSC_IDS = {'B5', 'B6', 'B7', 'B9', 'D9', 'D10', 'D11', 'D12', 'D13', 'D14', 'D15'};

% --- ANALYSIS PARAMETERS ---
R2_THRESHOLD = 0.99;        % Minimum R² for linear region
WINDOW_FRACTION = 0.1;      % Fraction of data points for sliding window
MIN_WINDOW = 5;             % Minimum window size

fprintf('SSP Biomechanics Analysis Pipeline\n');
fprintf('==================================\n\n');
fprintf('Data folder: %s\n', DATA_FOLDER);
fprintf('Results folder: %s\n\n', RESULTS_FOLDER);

%% Helper Functions

function [sampleID, condition, subgroup] = classifySample(filename, tflIDs, mscIDs)
    % CLASSIFYSAMPLE Parse filename to determine sample classification
    %
    % Inputs:
    %   filename - Name of the CSV file (e.g., 'SSP_2025-03-17_D1_NO.csv')
    %   tflIDs   - Cell array of TFL group subject IDs
    %   mscIDs   - Cell array of MSC group subject IDs
    %
    % Outputs:
    %   sampleID  - Subject identifier (e.g., 'D1')
    %   condition - 'NO' or 'OPER'
    %   subgroup  - 'NON', 'TFL', or 'MSC'
    
    % Remove .csv extension and split by underscore
    nameParts = strsplit(erase(filename, '.csv'), '_');
    
    % Determine condition
    if any(strcmp(nameParts, 'NO'))
        condition = 'NO';
    elseif any(strcmp(nameParts, 'OPER'))
        condition = 'OPER';
    else
        condition = 'Unknown';
    end
    
    % Find sample ID and determine subgroup
    sampleID = 'Unknown';
    subgroup = 'Unassigned';
    
    % Non-operated samples are always in NON group
    if strcmp(condition, 'NO')
        subgroup = 'NON';
    end
    
    % Check each part against known IDs
    for i = 1:length(nameParts)
        part = nameParts{i};
        if any(strcmp(tflIDs, part))
            sampleID = part;
            if strcmp(condition, 'OPER')
                subgroup = 'TFL';
            end
            break;
        elseif any(strcmp(mscIDs, part))
            sampleID = part;
            if strcmp(condition, 'OPER')
                subgroup = 'MSC';
            end
            break;
        elseif ~isempty(regexp(part, '^[BCD]\d{1,2}$', 'once'))
            % Fallback: match pattern like B1, C2, D15
            sampleID = part;
        end
    end
end

function [slope, intercept, r2, startIdx, endIdx] = findBestStiffness(x, y, windowSize, r2Threshold)
    % FINDBESTSTIFFNESS Find the stiffest linear region using sliding window
    %
    % Inputs:
    %   x           - Displacement values (mm)
    %   y           - Load values (N)
    %   windowSize  - Number of points in sliding window
    %   r2Threshold - Minimum R² to consider region as linear
    %
    % Outputs:
    %   slope     - Stiffness (N/mm)
    %   intercept - Y-intercept of linear fit
    %   r2        - R² of the best fit
    %   startIdx  - Start index of linear region
    %   endIdx    - End index of linear region
    
    nPoints = length(x);
    
    % Initialize outputs
    slope = NaN;
    intercept = NaN;
    r2 = NaN;
    startIdx = 0;
    endIdx = 0;
    
    if nPoints < windowSize
        warning('Insufficient data points (%d) for window size (%d)', nPoints, windowSize);
        return;
    end
    
    bestSlope = 0;
    bestR2 = -Inf;
    
    % Sliding window loop
    for i = 1:(nPoints - windowSize + 1)
        idx = i:(i + windowSize - 1);
        xWin = x(idx);
        yWin = y(idx);
        
        % Linear regression
        p = polyfit(xWin, yWin, 1);
        yPred = polyval(p, xWin);
        
        % Calculate R²
        ssRes = sum((yWin - yPred).^2);
        ssTot = sum((yWin - mean(yWin)).^2);
        if ssTot == 0
            r2Win = 0;
        else
            r2Win = 1 - (ssRes / ssTot);
        end
        
        currentSlope = p(1);
        
        % Criteria: highest slope with R² above threshold
        if r2Win >= r2Threshold
            if currentSlope > bestSlope
                bestSlope = currentSlope;
                slope = currentSlope;
                intercept = p(2);
                r2 = r2Win;
                startIdx = i;
                endIdx = i + windowSize - 1;
            end
        end
        
        % Fallback: track best R² if no region meets threshold
        if r2Win > bestR2 && bestSlope == 0
            bestR2 = r2Win;
            slope = currentSlope;
            intercept = p(2);
            r2 = r2Win;
            startIdx = i;
            endIdx = i + windowSize - 1;
        end
    end
end

%% Process All Files

% Get list of CSV files
files = dir(fullfile(DATA_FOLDER, '*.csv'));
nFiles = length(files);

fprintf('Found %d CSV files to process.\n\n', nFiles);

% Initialize results table
results = table();

for f = 1:nFiles
    filename = files(f).name;
    filepath = fullfile(DATA_FOLDER, filename);
    
    try
        % Classify sample
        [sampleID, condition, subgroup] = classifySample(filename, TFL_IDS, MSC_IDS);
        
        if strcmp(condition, 'Unknown')
            fprintf('[!] Skipping %s: Could not determine NO/OPER.\n', filename);
            continue;
        end
        
        % Read data
        data = readtable(filepath);
        
        % Handle different column names
        if ismember('LoadN', data.Properties.VariableNames)
            load_N = data.LoadN;
        elseif ismember('LoadkN', data.Properties.VariableNames)
            load_N = data.LoadkN * 1000;
        else
            fprintf('[!] Skipping %s: No load column found.\n', filename);
            continue;
        end
        
        displacement_mm = data.Crossheadmm;
        
        % Truncate at max load (failure point)
        [maxLoad, maxIdx] = max(load_N);
        x = displacement_mm(1:maxIdx);
        y = load_N(1:maxIdx);
        
        % Calculate energy (area under curve)
        energy_mJ = trapz(x, y);
        
        % Calculate stiffness
        windowSize = max(MIN_WINDOW, round(length(x) * WINDOW_FRACTION));
        [stiffness, ~, r2Score, startIdx, endIdx] = findBestStiffness(x, y, windowSize, R2_THRESHOLD);
        
        % Add to results
        newRow = table(...
            {filename}, {sampleID}, {subgroup}, maxLoad, stiffness, energy_mJ, r2Score, startIdx, endIdx, ...
            'VariableNames', {'Filename', 'SampleID', 'Subgroup', 'MaxLoad_N', 'Stiffness_N_mm', ...
                             'Energy_mJ', 'R2_Score', 'Linear_Start_Idx', 'Linear_End_Idx'});
        results = [results; newRow];
        
        fprintf('[OK] %s -> %s: MaxLoad=%.1fN, Stiffness=%.1fN/mm\n', ...
                filename, subgroup, maxLoad, stiffness);
        
    catch ME
        fprintf('[X] Error processing %s: %s\n', filename, ME.message);
    end
end

fprintf('\nProcessed %d samples successfully.\n', height(results));

%% Display Results

disp(' ');
disp('=== Sample Results ===');
disp(results);

%% Generate Group Statistics

groups = unique(results.Subgroup);
nGroups = length(groups);

fprintf('\n=== Group Statistics ===\n\n');

statsTable = table();

for g = 1:nGroups
    groupName = groups{g};
    groupData = results(strcmp(results.Subgroup, groupName), :);
    n = height(groupData);
    
    statsRow = table(...
        {groupName}, n, ...
        mean(groupData.MaxLoad_N), std(groupData.MaxLoad_N), ...
        mean(groupData.Stiffness_N_mm), std(groupData.Stiffness_N_mm), ...
        mean(groupData.Energy_mJ), std(groupData.Energy_mJ), ...
        'VariableNames', {'Subgroup', 'Count', ...
                         'MaxLoad_Mean', 'MaxLoad_Std', ...
                         'Stiffness_Mean', 'Stiffness_Std', ...
                         'Energy_Mean', 'Energy_Std'});
    statsTable = [statsTable; statsRow];
end

disp(statsTable);

%% Save Results to CSV

resultsPath = fullfile(RESULTS_FOLDER, 'Experiment_Master_Log_Detailed.csv');
writetable(results, resultsPath);
fprintf('\nSaved: %s\n', resultsPath);

statsPath = fullfile(RESULTS_FOLDER, 'Group_Statistics_Detailed.csv');
writetable(statsTable, statsPath);
fprintf('Saved: %s\n', statsPath);

%% Visualization: Group Comparison Bar Charts

figure('Position', [100, 100, 1400, 400]);

groupOrder = {'NON', 'TFL', 'MSC'};
colors = [0.18, 0.80, 0.44;   % Green for NON
          0.20, 0.60, 0.86;   % Blue for TFL
          0.91, 0.30, 0.24];  % Red for MSC

metrics = {'MaxLoad_N', 'Stiffness_N_mm', 'Energy_mJ'};
titles = {'Max Load (N)', 'Stiffness (N/mm)', 'Energy (mJ)'};

for m = 1:3
    subplot(1, 3, m);
    
    means = zeros(1, 3);
    stds = zeros(1, 3);
    counts = zeros(1, 3);
    
    for g = 1:3
        groupData = results(strcmp(results.Subgroup, groupOrder{g}), :);
        counts(g) = height(groupData);
        if counts(g) > 0
            means(g) = mean(groupData.(metrics{m}));
            stds(g) = std(groupData.(metrics{m}));
        end
    end
    
    b = bar(means, 'FaceColor', 'flat', 'EdgeColor', 'k', 'LineWidth', 1);
    b.CData = colors;
    
    hold on;
    errorbar(1:3, means, stds, 'k', 'LineStyle', 'none', 'LineWidth', 1.5, 'CapSize', 8);
    
    % Add sample counts
    for g = 1:3
        text(g, means(g) + stds(g) + max(means)*0.05, sprintf('n=%d', counts(g)), ...
             'HorizontalAlignment', 'center', 'FontSize', 10);
    end
    
    set(gca, 'XTickLabel', groupOrder, 'FontSize', 11);
    xlabel('Treatment Group', 'FontSize', 12);
    ylabel(titles{m}, 'FontSize', 12);
    title(titles{m}, 'FontSize', 14, 'FontWeight', 'bold');
    grid on;
    box on;
end

sgtitle('SSP Biomechanics: Group Comparison', 'FontSize', 16, 'FontWeight', 'bold');

% Save figure
plotPath = fullfile(RESULTS_FOLDER, 'Combined_Group_Plots.png');
saveas(gcf, plotPath);
fprintf('Saved: %s\n', plotPath);

%% Visualization: Load-Displacement Curves by Group

figure('Position', [100, 100, 1600, 500]);

for g = 1:3
    subplot(1, 3, g);
    groupName = groupOrder{g};
    groupData = results(strcmp(results.Subgroup, groupName), :);
    nSamples = height(groupData);
    
    % Color map
    if strcmp(groupName, 'NON')
        cmap = parula(max(nSamples, 1));
    elseif strcmp(groupName, 'TFL')
        cmap = cool(max(nSamples, 1));
    else
        cmap = hot(max(nSamples, 1));
    end
    
    legendEntries = {};
    
    for s = 1:nSamples
        filename = groupData.Filename{s};
        filepath = fullfile(DATA_FOLDER, filename);
        
        data = readtable(filepath);
        
        if ismember('LoadN', data.Properties.VariableNames)
            load_N = data.LoadN;
        else
            load_N = data.LoadkN * 1000;
        end
        
        displacement_mm = data.Crossheadmm;
        
        % Truncate at max load
        [~, maxIdx] = max(load_N);
        x = displacement_mm(1:maxIdx);
        y = load_N(1:maxIdx);
        
        plot(x, y, 'Color', cmap(s, :), 'LineWidth', 1.5);
        hold on;
        
        legendEntries{end+1} = groupData.SampleID{s};
    end
    
    xlabel('Displacement (mm)', 'FontSize', 12);
    ylabel('Load (N)', 'FontSize', 12);
    title(sprintf('%s Group (n=%d)', groupName, nSamples), 'FontSize', 14, 'FontWeight', 'bold');
    legend(legendEntries, 'Location', 'northwest', 'FontSize', 8);
    grid on;
    box on;
end

sgtitle('Load-Displacement Curves by Treatment Group', 'FontSize', 16, 'FontWeight', 'bold');

% Save individual group plots
for g = 1:3
    figure('Position', [100, 100, 800, 600]);
    groupName = groupOrder{g};
    groupData = results(strcmp(results.Subgroup, groupName), :);
    nSamples = height(groupData);
    
    if strcmp(groupName, 'NON')
        cmap = parula(max(nSamples, 1));
    elseif strcmp(groupName, 'TFL')
        cmap = cool(max(nSamples, 1));
    else
        cmap = hot(max(nSamples, 1));
    end
    
    legendEntries = {};
    
    for s = 1:nSamples
        filename = groupData.Filename{s};
        filepath = fullfile(DATA_FOLDER, filename);
        
        data = readtable(filepath);
        
        if ismember('LoadN', data.Properties.VariableNames)
            load_N = data.LoadN;
        else
            load_N = data.LoadkN * 1000;
        end
        
        displacement_mm = data.Crossheadmm;
        
        [~, maxIdx] = max(load_N);
        x = displacement_mm(1:maxIdx);
        y = load_N(1:maxIdx);
        
        plot(x, y, 'Color', cmap(s, :), 'LineWidth', 1.5);
        hold on;
        
        legendEntries{end+1} = groupData.SampleID{s};
    end
    
    xlabel('Displacement (mm)', 'FontSize', 12);
    ylabel('Load (N)', 'FontSize', 12);
    title(sprintf('%s Group (n=%d)', groupName, nSamples), 'FontSize', 14, 'FontWeight', 'bold');
    legend(legendEntries, 'Location', 'northwest', 'FontSize', 9);
    grid on;
    box on;
    
    plotPath = fullfile(RESULTS_FOLDER, sprintf('Plot_%s.png', groupName));
    saveas(gcf, plotPath);
    fprintf('Saved: %s\n', plotPath);
    close(gcf);
end

%% Visualization: Stiffness Region Example

figure('Position', [100, 100, 1000, 600]);

% Pick one sample from each group to show
for g = 1:3
    subplot(1, 3, g);
    groupName = groupOrder{g};
    groupData = results(strcmp(results.Subgroup, groupName), :);
    
    if height(groupData) == 0
        continue;
    end
    
    % Get first sample
    sampleRow = groupData(1, :);
    filename = sampleRow.Filename{1};
    filepath = fullfile(DATA_FOLDER, filename);
    
    data = readtable(filepath);
    
    if ismember('LoadN', data.Properties.VariableNames)
        load_N = data.LoadN;
    else
        load_N = data.LoadkN * 1000;
    end
    
    displacement_mm = data.Crossheadmm;
    
    [~, maxIdx] = max(load_N);
    x = displacement_mm(1:maxIdx);
    y = load_N(1:maxIdx);
    
    startIdx = sampleRow.Linear_Start_Idx;
    endIdx = sampleRow.Linear_End_Idx;
    
    % Plot full curve
    plot(x, y, 'b-', 'LineWidth', 2, 'DisplayName', 'Load-Displacement');
    hold on;
    
    % Highlight linear region
    if startIdx > 0 && endIdx > 0 && endIdx <= length(x)
        plot(x(startIdx:endIdx), y(startIdx:endIdx), 'r-', 'LineWidth', 3, ...
             'DisplayName', sprintf('Linear Region (R^2=%.4f)', sampleRow.R2_Score));
    end
    
    xlabel('Displacement (mm)', 'FontSize', 11);
    ylabel('Load (N)', 'FontSize', 11);
    title(sprintf('%s: %s\nStiffness=%.1f N/mm', groupName, sampleRow.SampleID{1}, ...
                  sampleRow.Stiffness_N_mm), 'FontSize', 12);
    legend('Location', 'northwest');
    grid on;
    box on;
end

sgtitle('Stiffness Calculation: Linear Region Examples', 'FontSize', 14, 'FontWeight', 'bold');

plotPath = fullfile(RESULTS_FOLDER, 'Stiffness_Region_Examples.png');
saveas(gcf, plotPath);
fprintf('Saved: %s\n', plotPath);

%% Summary

fprintf('\n');
fprintf('============================================\n');
fprintf('ANALYSIS COMPLETE\n');
fprintf('============================================\n');
fprintf('\nTotal samples processed: %d\n', height(results));
fprintf('Results saved to: %s\n', RESULTS_FOLDER);
fprintf('\nFiles generated:\n');
fprintf('  - Experiment_Master_Log_Detailed.csv\n');
fprintf('  - Group_Statistics_Detailed.csv\n');
fprintf('  - Combined_Group_Plots.png\n');
fprintf('  - Plot_NON.png, Plot_TFL.png, Plot_MSC.png\n');
fprintf('  - Stiffness_Region_Examples.png\n');
